### ISSUE TITLE
Compatibility Issues

### ISSUE THREAD

**Issue by paulofaria** [open-swift/docs#8]
Open Swift's goal is to solve cross-project compatibility. But we have some issues in the way we currently handle things. Specifically:

```swift
// Open Swift
struct Request { ... }

// Framework A
extension Request {
  var cookie: Cookie { ... }
}

// Framework B
extension Request {
  var cookie: Cookie { ... }
}

// Application
import FrameworkA
import FrameworkB

let cookie = request.cookie // trouble
```

We should discuss a solution for this. Some that come to mind:

- Add a prefix to any computed property defined in an extension so we don't have name clashes.

Example:

```swift
extension Request {
  var fwkCookie: Cookie { ... }
}
```

- Create a proposal for Swift to allow disambiguation.

Example:

```swift
import FrameworkA
import FrameworkB

let cookie: FrameworkA.Cookie = request.cookie
```

- Make Open Swift a self-contained solution that doesn't need extensions (in general). In other words, change from protocol standards to a standard library.

Please leave any other suggestions in the comments and I'll update the issue description. Or update it directly if you're part of the organization.

**Comment by bradhilton** [comment]
Please no prefixes! Let's find another solution.

**Comment by Joannis** [comment]
I refuse to use prefixes. That's going to cause a lot of ugliness.

And I think it's worth a proposal towards Swift since this is a great solution.
But disambiguating the code manually isn't a bad solution either.

**Comment by paulofaria** [comment]
I don't like the manual disambiguation cause we would be placing the burden on the consumer when we're supposed to be trying to make his life easier.

**Comment by Joannis** [comment]
I agree that it's a burden.. but we can't cover all cases I think.

**Comment by paulofaria** [comment]
The problem is that instead of us doing the ugly prefixes we would be making the consumer do it himself, haha. I don't think we should that to them. :(

**Comment by Joannis** [comment]
I disagree. Either you add prefixes where users use two libraries with the same variable in an extension. Or you add prefixes for everyone. I'd rather focus on the large userbase rather than the smaller one.

**Comment by paulofaria** [comment]
@Joannis no! I'm against prefixes for everyone! I think we can find a better solution.

**Comment by tannernelson** [comment]
I hope the Swift compiler guys can come up with a good solution for this, but as it stands I think it would make sense to move what is necessary to Open Swift. It's hard to define what is "necessary" but I think that we could safely say for now that what is in an official HTTP spec (RFC, etc) should be protocolized in Open Swift. It doesn't make sense for `S4.Middleware` to be incompatible between two frameworks when it decides to use cookies. These kinds of fundamentals need to be covered.

That being said, I don't want to stifle creativity or limit room for innovation in the implementations. But the RFC specs are the specs, and if we conform to those as well as we can, we should be able to provide cross project compatibility without putting limits on the implementations.

**Comment by Joannis** [comment]
What about importing a specific class or struct?

`import struct BSON.Document` is completely valid

`import importKind module.symbolName`
`import module.submodule`

This is from: “The Swift Programming Language (Swift 2.1).” ( https://itun.es/nl/jEUH0.l )

**Comment by dreymonde** [comment]
Actually, that’s exactly what I was talking about here https://github.com/open-swift/C7/pull/37#issuecomment-219969248
The main problem is, I suppose, not the compiler-or-whatever issues, but the general inconsistency we’re bringing to the community, relying heavily on extensions by other frameworks. It’s completely understandable that high-level stuff will look and work differently in different solution, but for things like cookies, HTTP requests, JSON and so on — I think there should be consistency, so the developer will expect the same reasonable behavior using different frameworks (and that is especially important when combining those frameworks!). 
So I’m talking that we should do it not because of ambiguity issues (or at least not *only* because of ambiguity issues).

**Comment by Joannis** [comment]
@dreymonde I completely concur. It's at least worth a proper reconsideration

**Comment by czechboy0** [comment]
Agreed, if two frameworks a) create the same property of b) with the same intention that c) behaves pretty much identically, that might be a signal to move it up the stack to Open Swift. After all, if someone wants to add `crazyCookies` on top of their framework that do something else, they can, but good defaults are very important.

**Comment by paulofaria** [comment]
Yeah! I think some things could still be protocols like HTTP parser and serializer. Cause sometimes you want a full fledged parser that is very strict and all, but sometimes you're gonna reverse proxy with nginx and all you need is a lightweight naive parser. But in general, HTTP headers and media types are something you can't get too crazy about. And if you really want to, you can still add extensions, like the `crazyCookies` @czechboy0 mentioned.
